File: backend/Huffman.java
import java.io.*;
import java.util.PriorityQueue;
import java.util.HashMap;
import java.util.Map;

class HuffmanNode implements Comparable<HuffmanNode> {
    int freq;
    int c; // Use int to handle 0-255 byte values, -1 for internal node
    HuffmanNode left;
    HuffmanNode right;

    public HuffmanNode(int c, int freq) {
        this.c = c;
        this.freq = freq;
        this.left = null;
        this.right = null;
    }

    @Override
    public int compareTo(HuffmanNode node) {
        return this.freq - node.freq;
    }
}

public class Huffman {

    public static void main(String[] args) {
        if (args.length < 3) {
            System.err.println("Usage: java Huffman <compress|decompress> <inputFile> <outputFile>");
            System.exit(1);
        }

        String action = args[0];
        String inputFile = args[1];
        String outputFile = args[2];

        try {
            if ("compress".equals(action)) {
                compress(inputFile, outputFile);
            } else if ("decompress".equals(action)) {
                decompress(inputFile, outputFile);
            } else {
                System.err.println("Invalid action. Use 'compress' or 'decompress'.");
                System.exit(1);
            }
        } catch (IOException e) {
            e.printStackTrace();
            System.exit(1);
        }
    }

    private static void compress(String inputFile, String outputFile) throws IOException {
        File file = new File(inputFile);
        byte[] data = new byte[(int) file.length()];
        try (FileInputStream fis = new FileInputStream(file)) {
            fis.read(data);
        }
        
        if (data.length == 0) {
            new FileOutputStream(outputFile).close();
            return;
        }

        int[] freqs = new int[256];
        for (byte b : data) {
            freqs[b & 0xFF]++;
        }

        HuffmanNode root = buildTree(freqs);
        String[] codes = new String[256];
        buildCodes(root, "", codes);

        int totalBits = 0;
        for (int i = 0; i < 256; i++) {
            if (codes[i] != null) {
                totalBits += freqs[i] * codes[i].length();
            }
        }

        int padding = (8 - (totalBits % 8)) % 8;

        try (DataOutputStream dos = new DataOutputStream(new FileOutputStream(outputFile))) {
            // Write header
            for (int i = 0; i < 256; i++) {
                dos.writeInt(freqs[i]);
            }
            dos.writeByte(padding);

            // Write packed bits
            int currentByte = 0;
            int numBits = 0;
            for (byte b : data) {
                String code = codes[b & 0xFF];
                for (int i = 0; i < code.length(); i++) {
                    currentByte = (currentByte << 1) | (code.charAt(i) - '0');
                    numBits++;
                    if (numBits == 8) {
                        dos.write(currentByte);
                        currentByte = 0;
                        numBits = 0;
                    }
                }
            }
            
            // Write remaining bits with padding
            if (numBits > 0) {
                currentByte = currentByte << (8 - numBits);
                dos.write(currentByte);
            }
        }
    }

    private static void decompress(String inputFile, String outputFile) throws IOException {
        File file = new File(inputFile);
        if (file.length() == 0) {
            new FileOutputStream(outputFile).close();
            return;
        }

        try (DataInputStream dis = new DataInputStream(new FileInputStream(file));
             FileOutputStream fos = new FileOutputStream(outputFile)) {

            int[] freqs = new int[256];
            long totalOriginalBytes = 0;
            for (int i = 0; i < 256; i++) {
                freqs[i] = dis.readInt();
                totalOriginalBytes += freqs[i];
            }
            int padding = dis.readByte();

            HuffmanNode root = buildTree(freqs);
            if (root == null) return;

            // Handle edge case of single repeating character
            if (root.left == null && root.right == null) {
                for (long i = 0; i < totalOriginalBytes; i++) {
                    fos.write(root.c);
                }
                return;
            }

            long outPos = 0;
            HuffmanNode curr = root;

            byte[] buffer = new byte[8192];
            int read;
            while ((read = dis.read(buffer)) != -1) {
                for (int b = 0; b < read; b++) {
                    int currentByte = buffer[b] & 0xFF;
                    int bitsToProcess = 8;
                    
                    // On the very last byte, account for padding
                    if (dis.available() == 0 && b == read - 1) {
                        bitsToProcess = 8 - padding;
                    }

                    for (int i = 0; i < bitsToProcess && outPos < totalOriginalBytes; i++) {
                        int bit = (currentByte >> (7 - i)) & 1;
                        curr = (bit == 0) ? curr.left : curr.right;

                        if (curr.left == null && curr.right == null) {
                            fos.write(curr.c);
                            outPos++;
                            curr = root;
                        }
                    }
                }
            }
        }
    }

    private static HuffmanNode buildTree(int[] charFreq) {
        PriorityQueue<HuffmanNode> pq = new PriorityQueue<>();
        for (int i = 0; i < 256; i++) {
            if (charFreq[i] > 0) {
                pq.add(new HuffmanNode(i, charFreq[i]));
            }
        }

        if (pq.isEmpty()) return null;
        if (pq.size() == 1) {
            HuffmanNode leaf = pq.poll();
            HuffmanNode root = new HuffmanNode(-1, leaf.freq);
            root.left = leaf; // Forces at least one bit per character
            return root;
        }

        while (pq.size() > 1) {
            HuffmanNode left = pq.poll();
            HuffmanNode right = pq.poll();
            HuffmanNode parent = new HuffmanNode(-1, left.freq + right.freq);
            parent.left = left;
            parent.right = right;
            pq.add(parent);
        }

        return pq.poll();
    }

    private static void buildCodes(HuffmanNode root, String prefix, String[] codes) {
        if (root == null) return;
        if (root.left == null && root.right == null && root.c != -1) {
            codes[root.c] = prefix.isEmpty() ? "0" : prefix;
            return;
        }
        buildCodes(root.left, prefix + "0", codes);
        buildCodes(root.right, prefix + "1", codes);
    }
}

File: backend/huffman.js
class Node {
  constructor(char, freq, left = null, right = null) {
    this.char = char;
    this.freq = freq;
    this.left = left;
    this.right = right;
  }
}

function buildFreqs(buffer) {
  const freqs = new Array(256).fill(0);
  for (let i = 0; i < buffer.length; i++) freqs[buffer[i]]++;
  return freqs;
}

function buildTree(freqs) {
  const pq = [];
  for (let i = 0; i < 256; i++) {
    if (freqs[i] > 0) pq.push(new Node(i, freqs[i]));
  }
  if (pq.length === 0) return null;
  pq.sort((a, b) => a.freq - b.freq);

  while (pq.length > 1) {
    const left = pq.shift();
    const right = pq.shift();
    const parent = new Node(null, left.freq + right.freq, left, right);
    let inserted = false;
    for (let i = 0; i < pq.length; i++) {
      if (pq[i].freq > parent.freq) { pq.splice(i, 0, parent); inserted = true; break; }
    }
    if (!inserted) pq.push(parent);
  }
  return pq[0];
}

function buildCodes(node, prefix = "", codes = {}) {
  if (!node) return codes;
  if (node.char !== null) { codes[node.char] = prefix || "0"; return codes; }
  buildCodes(node.left, prefix + "0", codes);
  buildCodes(node.right, prefix + "1", codes);
  return codes;
}

function compress(buffer) {
  if (buffer.length === 0) return Buffer.alloc(0);
  const freqs = buildFreqs(buffer);
  const tree = buildTree(freqs);
  const codes = buildCodes(tree);

  let totalBits = 0;
  for (let i = 0; i < 256; i++) totalBits += freqs[i] * (codes[i] ? codes[i].length : 0);
  const padding = (8 - (totalBits % 8)) % 8;
  const numBytes = Math.ceil(totalBits / 8);
  const dataBuffer = Buffer.alloc(numBytes);

  let bitPos = 0;
  for (let i = 0; i < buffer.length; i++) {
    const code = codes[buffer[i]];
    for (let j = 0; j < code.length; j++) {
      if (code[j] === "1") dataBuffer[Math.floor(bitPos / 8)] |= (1 << (7 - (bitPos % 8)));
      bitPos++;
    }
  }

  const header = Buffer.alloc(256 * 4 + 1);
  for (let i = 0; i < 256; i++) header.writeUInt32LE(freqs[i], i * 4);
  header.writeUInt8(padding, 256 * 4);
  return Buffer.concat([header, dataBuffer]);
}

function decompress(buffer) {
  if (buffer.length === 0) return Buffer.alloc(0);
  const headerSize = 256 * 4 + 1;
  const freqs = new Array(256).fill(0);
  let totalOriginalBytes = 0;
  for (let i = 0; i < 256; i++) { freqs[i] = buffer.readUInt32LE(i * 4); totalOriginalBytes += freqs[i]; }
  const padding = buffer.readUInt8(256 * 4);
  const data = buffer.subarray(headerSize);
  const tree = buildTree(freqs);
  if (!tree) return Buffer.alloc(0);

  const outBuffer = Buffer.allocUnsafe(totalOriginalBytes);
  let outPos = 0;
  let curr = tree;

  if (curr.char !== null) {
    for (let i = 0; i < totalOriginalBytes; i++) outBuffer[i] = curr.char;
    return outBuffer;
  }

  const totalBits = data.length * 8 - padding;
  let bitPos = 0;
  while (bitPos < totalBits && outPos < totalOriginalBytes) {
    const bit = (data[Math.floor(bitPos / 8)] >> (7 - (bitPos % 8))) & 1;
    curr = bit === 0 ? curr.left : curr.right;
    if (curr.char !== null) { outBuffer[outPos++] = curr.char; curr = tree; }
    bitPos++;
  }
  return outBuffer;
}

module.exports = { compress, decompress };
